python/fusion_engine_client/parsers/mixed_log_reader.py

```diff
@@ -107,6 +111,18 @@ def __init__(self, input_file, warn_on_gaps: bool = False, show_progress: bool =
         self._original_index = fast_indexer.fast_generate_index(input_path, force_reindex=ignore_index,
                                                                 save_index=save_index, max_bytes=max_bytes)
         self.next_index_elem = 0
+        self.index = self._original_index
+        self.filtered_message_types = False
+
+        # Now, apply filters -> apply pose message filter, then look at available message types
```

Comment:
What's up with this comment? Does this match the style of other comments in this repo? What does it mean?

Reply:
This comment doesn't really match the code. I'm not totally sure what this sentence is saying, so it needs clarification, and there is no `populate_xx()` function?

I think this could be better said as like:

> Before restricting to the user-requested set of source IDs, first look at data in the file to determine the list of all available source IDs (within the user's time range).

This step should apply the time range filter but _not_ the message type filter, then call `get_available_source_ids()` to generate the list of IDs, which will be stored in `self.available_source_ids` (which should be a `set`).

Next, step 2 if the user specified the `source_ids` argument here (stored in `self.requested_source_ids`), I would update `self.available_source_ids` to be the intersection of the two. Later, in `read()`:
```py
if source_ids is None:
  source_ids = None self.requested_source_ids is None else self.available_source_ids
else:
  source_ids = self.available_source_ids.intersection(source_ids)
```
If `source_ids is not None` you apply it to the messages. Otherwise you can skip checking IDs altogether.

Finally, step 3 should clear the filters and reapply the time range, plus the message types (with a comment explaining it).

---

I would actually recommend you _do_ create a populate function to call here to make things a little more explicit and clear, but I would make it protected: `_populate_available_source_ids()`. Then `get_available_source_ids()` should simply return `self.available_source_ids`.

Reply:
Actually no, scratch most of that. I see how `self.requested_source_ids` is being used in `filter_in_place()`, and it is consistent with time range and message types.

We should still separate populate and get, but `self.available_source_ids` should simply be the list of all available source IDs regardless of time range or anything else.

We should make the argument default above `None` though, not `[0]`, and both `self.available_source_ids` and `self.requested_source_ids` should be converted to a `set()` for efficiency and convenience.